### ISSUE TITLE
Surprisingly neither hypothesis nor hypofuzz finds the failing example, where there is a small failing example

### ISSUE THREAD

**Issue by eivindjahren** [Zac-HD/hypofuzz#39]
The following test has the failing example `"^RUN"`, but neither running hypothesis over night nor running hypofuzz for a couple of hours finds the failing string: 
```python
from hypothesis import given
import hypothesis.strategies as st


magic_words = ["RUN", "JUMP"]


def parse_spells(output: str) -> dict[str, str]:
    spells = {}
    for line in output.splitlines():
        tokens = line.split("^")
        if len(tokens) == 2:
            name, spell = tokens
            if spell not in magic_words:
                continue
            spells[name] = spell
    return spells


@given(st.text())
def test_parse_spells(text):
    assert parse_spells(text) == {}
```
This is a simplified version of something we had in production.

**Comment by Zac-HD** [comment]
I'm not surprised - exact comparisons are very difficult to find by chance, and HypoFuzz doesn't (yet?) use any techniques to make this easier.  Options would include:

- constructing a "dictionary" of interesting constants, e.g. by pulling them out of the AST (as in https://github.com/Zac-HD/hypofuzz/issues/8).  To solve this particular case would take some fancy combinations too, though.
- rewriting comparison operators to give more detailed feedback, e.g. a "virtual branch" for each matching element and maybe a `target()` call for any length-mismatch.  Native-code fuzzers often split comparisons into byte-by-byte checks for this reason.  However, Unicode strings are much less friendly for this, since there are many many more possibilities per element.
- using an SMT solver, e.g. via [the experimental `crosshair` backend](https://hypothesis.readthedocs.io/en/latest/strategies.html#alternative-backends-for-hypothesis).  Stabilizing this and integrating it into HypoFuzz is one of my dreams for the ecosystem.

**Comment by eivindjahren** [comment]
Yes, a smt solver could potentially find it. I tried to see whether `hypothesis-crosshair` would do it, but I ran into an issue: https://github.com/pschanely/CrossHair/issues/263 .

The fuzzing dictionary would probably be interesting. I thought that in this case I have left what amounts to a very simple passphrase for the fuzzer to find. Creating a fuzzing dictionary by calling `strings` on the binary is in that case pretty much what [CeWL](https://www.geeksforgeeks.org/cewl-tool-creating-custom-wordlists-tool-in-kali-linux/) does for password cracking.

I have used the lark extension to create more interesting text sometimes:

```python
from hypothesis import given
from hypothesis.extra.lark import from_lark
import hypothesis.strategies as st

more_interesting = from_lark(...)

@given(st.one_of(st.text(), more_interesting))
def test_parse_spells(text):
    ...
```

**Comment by pschanely** [comment]
At least right now, CrossHair isn't amazing at string parsing tasks like this, but with native crosshair (and some expansion of default timeouts), it'll work; you should get a counterexample after a minute running it on crosshair-web, [here](https://crosshair-web.org/?crosshair=0.1&python=3.8&gist=221b880de834a069123b20cc2f50d396).

There's no obvious reason it shouldn't work with hypothesis too - I'll investigate and follow up in https://github.com/pschanely/CrossHair/issues/263 shortly.

**Comment by eivindjahren** [comment]
Thank you so much @pschanely ! Sorry if reported this in the wrong repository.

**Comment by Zac-HD** [comment]
https://github.com/HypothesisWorks/hypothesis/pull/4356 should improve our chances of finding such a bug!

I think it doesn't make much sense to put combinations-of-constants (-and-random) in Hypothesis, where there's not much time to explore anyway, but that might be a neat trick to include in the Hypofuzz backend (cf #36).  We'd also want to break some of the 'global constants' into smaller pieces I suppose.

**Comment by Zac-HD** [comment]
Closing this in favor of #8, which outlines my preferred solution.
